File: Code.java
//
// Definition for binary tree:
// class Tree<T> {
//   Tree(T x) {
//     value = x;
//   }
//   T value;
//   Tree<T> left;
//   Tree<T> right;
// }
int kthSmallestInBST(Tree<Integer> t, int k) {
    Tree<Integer> curr = t, pre = null;
    
    int count = 0;
    
    while(curr != null)
    {           
        if(curr.left == null)
        {
            count++;
            if(count == k) return curr.value;
            curr = curr.right;
        }
        else{
            pre = curr.left;
            while(pre.right != null && curr != pre.right)
                pre = pre.right;
            
            if(pre.right == null)
            {
                pre.right = curr;
                curr = curr.left;   
            }
            else{
                count++;
                pre.right = null;
                if(count == k) return curr.value;
                curr = curr.right;
            }
        }
        
            
    }
    
    return k;
}

File: Code.c
//
//
// Definition for binary tree:
// typedef struct tree_##name {
//   type value;
//   struct tree_##name *left;
//   struct tree_##name *right;
// } tree_##name;
//
// tree_##name *alloc_tree_##name(type v) {
//   tree_##name *t = calloc(1, sizeof(tree_##name));
//   t->value = v;
//   return t;
// }
int kthSmallestInBST(tree_integer * t, int k) {
    tree_integer * curr = t;
    tree_integer * prev;
    int count = 0;
    
    while(curr != NULL){
        if(curr->left == NULL){
            count++;
            if(count == k) return curr->value;
            curr = curr->right;
        }
        else{
            prev = curr->left;
            while(prev->right != NULL && curr != prev->right)
                prev = prev->right;
            
            if(prev->right == NULL){
                prev->right = curr;
                curr = curr->left;
            }else{
                count++;
                prev->right = NULL;
                if(count == k) return curr->value;
                curr = curr->right;
            }
        }
    }
    
    return k;
}
